File: web/tests/game-status.test.js
/**
 * Headless tests for web/js/lib/game-status.js (WP 4a.3).
 * Run: node --test "web/tests/*.test.js"   (see web/tests/README.md)
 */
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  KIND,
  findLiveJob,
  indexLiveJobsByAppid,
  hasVisibleCacheContent,
  hasProtectedCacheContent,
  dispKind,
  statusAction,
  isJobStateTransition,
} from "../js/lib/game-status.js";

const game = (over) => ({
  appid: 1,
  name: "Aurora Cascade",
  status: "idle",
  last_prefill_at: null,
  depot_count: 0,
  size_bytes: null,
  needs_force: false,
  ...over,
});

const job = (over) => ({
  id: 1,
  appid: 1,
  type: "prefill",
  status: "running",
  ...over,
});

// ---------------------------------------------------------------------
// findLiveJob / indexLiveJobsByAppid
// ---------------------------------------------------------------------

test("findLiveJob matches only prefill jobs in running/paused", () => {
  const jobs = [
    job({ id: 1, appid: 10, status: "running" }),
    job({ id: 2, appid: 20, status: "paused" }),
    job({ id: 3, appid: 30, status: "queued" }),
    job({ id: 4, appid: 40, status: "done" }),
    job({ id: 5, appid: 50, status: "running", type: "gc" }),
  ];
  assert.equal(findLiveJob(jobs, 10).id, 1);
  assert.equal(findLiveJob(jobs, 20).id, 2);
  assert.equal(findLiveJob(jobs, 30), undefined, "queued is not live (matches mockup jobFor)");
  assert.equal(findLiveJob(jobs, 40), undefined);
  assert.equal(findLiveJob(jobs, 50), undefined, "a GC job must never drive the download pill");
});

test("indexLiveJobsByAppid is a Map keyed by appid, same filtering as findLiveJob", () => {
  const jobs = [job({ id: 1, appid: 10, status: "running" }), job({ id: 2, appid: 20, status: "queued" })];
  const map = indexLiveJobsByAppid(jobs);
  assert.equal(map.size, 1);
  assert.equal(map.get(10).id, 1);
  assert.equal(map.get(20), undefined);
});

test("indexLiveJobsByAppid tolerates a null/undefined jobs snapshot", () => {
  assert.equal(indexLiveJobsByAppid(undefined).size, 0);
  assert.equal(indexLiveJobsByAppid(null).size, 0);
});

// ---------------------------------------------------------------------
// hasVisibleCacheContent / hasProtectedCacheContent
// ---------------------------------------------------------------------

test("hasVisibleCacheContent requires a positive numeric size_bytes", () => {
  assert.equal(hasVisibleCacheContent(game({ size_bytes: 100 })), true);
  assert.equal(hasVisibleCacheContent(game({ size_bytes: 0 })), false);
  assert.equal(hasVisibleCacheContent(game({ size_bytes: null })), false);
  assert.equal(hasVisibleCacheContent(game({ size_bytes: undefined })), false);
});

test("hasProtectedCacheContent mirrors the server predicate: idle + never-prefilled + no job => false", () => {
  assert.equal(
    hasProtectedCacheContent(game({ status: "idle", last_prefill_at: null }), false),
    false,
  );
});

test("hasProtectedCacheContent: any of status!=idle / last_prefill_at set / active job protects", () => {
  assert.equal(hasProtectedCacheContent(game({ status: "done" }), false), true);
  assert.equal(hasProtectedCacheContent(game({ status: "error" }), false), true);
  assert.equal(
    hasProtectedCacheContent(game({ status: "idle", last_prefill_at: "2026-08-01T00:00:00Z" }), false),
    true,
  );
  assert.equal(hasProtectedCacheContent(game({ status: "idle", last_prefill_at: null }), true), true);
});

test("hasVisibleCacheContent and hasProtectedCacheContent can disagree (the 'last cached remnant' case)", () => {
  // status 'done' but the depot behind it was just reclaimed as an
  // orphaned remnant by an UNRELATED delete — api/README.md "Last cached
  // remnants". The grid must show this as Not cached; the deletion-side
  // predicate must still treat it as protecting a depot until it is
  // re-prefilled (fail-closed).
  const remnant = game({ status: "done", size_bytes: null, last_prefill_at: "2026-08-01T00:00:00Z" });
  assert.equal(hasVisibleCacheContent(remnant), false);
  assert.equal(hasProtectedCacheContent(remnant, false), true);
});

// ---------------------------------------------------------------------
// dispKind
// ---------------------------------------------------------------------

test("dispKind: a live running job overrides cache state", () => {
  assert.equal(dispKind(game({ status: "done", size_bytes: 100 }), job({ status: "running" })), KIND.RUNNING);
});
test("dispKind: a live paused job overrides cache state", () => {
  assert.equal(dispKind(game({ status: "idle" }), job({ status: "paused" })), KIND.PAUSED);
});
test("dispKind: no live job, status error => ERROR regardless of bytes", () => {
  assert.equal(dispKind(game({ status: "error", size_bytes: 500 }), undefined), KIND.ERROR);
  assert.equal(dispKind(game({ status: "error", size_bytes: null }), undefined), KIND.ERROR);
});
test("dispKind: no live job, done + visible bytes => CACHED", () => {
  assert.equal(dispKind(game({ status: "done", size_bytes: 500 }), undefined), KIND.CACHED);
});
test("dispKind: no live job, done but zero/no bytes => NONE (invariant, mockup round 5 finding 6)", () => {
  assert.equal(dispKind(game({ status: "done", size_bytes: null }), undefined), KIND.NONE);
  assert.equal(dispKind(game({ status: "done", size_bytes: 0 }), undefined), KIND.NONE);
});
test("dispKind: idle, no bytes => NONE", () => {
  assert.equal(dispKind(game({ status: "idle", size_bytes: null }), undefined), KIND.NONE);
});

// ---------------------------------------------------------------------
// statusAction
// ---------------------------------------------------------------------

test("statusAction: null while multi-select is active, regardless of state", () => {
  assert.equal(statusAction(game({ status: "idle" }), undefined, true), null);
  assert.equal(statusAction(game({ status: "done", size_bytes: 5 }), job({ status: "running" }), true), null);
});
test("statusAction: running job => pause", () => {
  const a = statusAction(game(), job({ status: "running" }), false);
  assert.equal(a.type, "pause");
});
test("statusAction: paused job => resume", () => {
  const a = statusAction(game(), job({ status: "paused" }), false);
  assert.equal(a.type, "resume");
});
test("statusAction: not-cached game => download", () => {
  const a = statusAction(game({ status: "idle", size_bytes: null }), undefined, false);
  assert.equal(a.type, "download");
  assert.equal(a.title, "Download to cache");
});
test("statusAction: errored game => download, titled as a retry", () => {
  const a = statusAction(game({ status: "error" }), undefined, false);
  assert.equal(a.type, "download");
  assert.equal(a.title, "Retry download");
});
test("statusAction: a cached game is inert (never a silent re-download)", () => {
  assert.equal(statusAction(game({ status: "done", size_bytes: 5 }), undefined, false), null);
});

// ---------------------------------------------------------------------
// isJobStateTransition — the round-7 "don't rebuild on a no-op tick" guard
// ---------------------------------------------------------------------

test("isJobStateTransition: same status (even with a grown log_excerpt) is NOT a transition", () => {
  const a = job({ status: "running", log_excerpt: "line 1" });
  const b = job({ status: "running", log_excerpt: "line 1\nline 2\nline 3" });
  assert.equal(isJobStateTransition(a, b), false);
});
test("isJobStateTransition: a status change IS a transition", () => {
  assert.equal(isJobStateTransition(job({ status: "running" }), job({ status: "paused" })), true);
});
test("isJobStateTransition: a brand-new job (no prev) is always a transition", () => {
  assert.equal(isJobStateTransition(undefined, job({ status: "running" })), true);
});
test("isJobStateTransition: a job disappearing (no curr) is always a transition", () => {
  assert.equal(isJobStateTransition(job({ status: "running" }), undefined), true);
});

File: app/app/src/test/java/dev/steamvault/app/ui/library/logic/GameStatusTest.kt
package dev.steamvault.app.ui.library.logic

import dev.steamvault.app.net.model.GameSummary
import dev.steamvault.app.net.model.JobSummary
import dev.steamvault.app.ui.status.StatusKind
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

private fun game(
    appid: Int = 1,
    status: String = "idle",
    lastPrefillAt: String? = null,
    sizeBytes: Long? = null,
    depotCount: Int = 0,
): GameSummary = GameSummary(
    appid = appid,
    name = "Game $appid",
    status = status,
    last_prefill_at = lastPrefillAt,
    last_manifest_check = null,
    depot_count = depotCount,
    size_bytes = sizeBytes,
    needs_force = false,
)

private fun job(
    appid: Int = 1,
    type: String = "prefill",
    status: String = "running",
    id: Int = 1,
): JobSummary = JobSummary(
    id = id,
    appid = appid,
    type = type,
    status = status,
    created_at = "2026-08-01T00:00:00Z",
)

class GameStatusTest {

    // ---- findLiveJob / indexLiveJobsByAppid ----------------------------

    @Test
    fun `findLiveJob matches a running prefill job for the appid`() {
        val jobs = listOf(job(appid = 42, status = "running"))
        assertEquals(jobs[0], findLiveJob(jobs, 42))
    }

    @Test
    fun `findLiveJob ignores a queued job`() {
        val jobs = listOf(job(appid = 42, status = "queued"))
        assertNull(findLiveJob(jobs, 42))
    }

    @Test
    fun `findLiveJob ignores a gc job even if running`() {
        val jobs = listOf(job(appid = 42, type = "gc", status = "running"))
        assertNull(findLiveJob(jobs, 42))
    }

    @Test
    fun `findLiveJob ignores a different appid`() {
        val jobs = listOf(job(appid = 1, status = "running"))
        assertNull(findLiveJob(jobs, 2))
    }

    @Test
    fun `indexLiveJobsByAppid only keeps live prefill jobs`() {
        val jobs = listOf(
            job(appid = 1, status = "running"),
            job(appid = 2, status = "queued"),
            job(appid = 3, type = "gc", status = "running"),
            job(appid = 4, status = "paused"),
        )
        val index = indexLiveJobsByAppid(jobs)
        assertEquals(setOf(1, 4), index.keys)
    }

    // ---- hasVisibleCacheContent / hasProtectedCacheContent -------------

    @Test
    fun `hasVisibleCacheContent requires positive size_bytes`() {
        assertFalse(hasVisibleCacheContent(game(sizeBytes = null)))
        assertFalse(hasVisibleCacheContent(game(sizeBytes = 0)))
        assertTrue(hasVisibleCacheContent(game(sizeBytes = 1)))
    }

    @Test
    fun `hasProtectedCacheContent is false only for idle plus never-prefilled plus no active job`() {
        assertFalse(hasProtectedCacheContent("idle", null, hasActiveJob = false))
        assertTrue(hasProtectedCacheContent("idle", null, hasActiveJob = true))
        assertTrue(hasProtectedCacheContent("idle", "2026-08-01T00:00:00Z", hasActiveJob = false))
        assertTrue(hasProtectedCacheContent("done", null, hasActiveJob = false))
        assertTrue(hasProtectedCacheContent("error", null, hasActiveJob = false))
    }

    // ---- dispKind -------------------------------------------------------

    @Test
    fun `dispKind a live running job overrides everything`() {
        val g = game(status = "error", sizeBytes = 5)
        assertEquals(StatusKind.RUNNING, dispKind(g, job(status = "running")))
    }

    @Test
    fun `dispKind a live paused job renders paused`() {
        val g = game(status = "done", sizeBytes = 5)
        assertEquals(StatusKind.PAUSED, dispKind(g, job(status = "paused")))
    }

    @Test
    fun `dispKind error status with no live job is error`() {
        assertEquals(StatusKind.ERROR, dispKind(game(status = "error"), null))
    }

    @Test
    fun `dispKind cached requires visible bytes -- a done status with null size is NOT cached`() {
        // Cache-content invariant (mockup round 5, finding 6): a "last
        // cached remnant" done row with size_bytes=null must render NONE.
        assertEquals(StatusKind.NONE, dispKind(game(status = "done", sizeBytes = null), null))
    }

    @Test
    fun `dispKind cached with real bytes`() {
        assertEquals(StatusKind.CACHED, dispKind(game(status = "done", sizeBytes = 5), null))
    }

    @Test
    fun `dispKind never-prefilled idle game is none`() {
        assertEquals(StatusKind.NONE, dispKind(game(status = "idle", sizeBytes = null), null))
    }

    // ---- statusAction -----------------------------------------------------

    @Test
    fun `statusAction is null while selecting, regardless of state`() {
        assertNull(statusAction(game(status = "idle"), null, selecting = true))
        assertNull(statusAction(game(status = "error"), null, selecting = true))
        assertNull(statusAction(game(status = "done", sizeBytes = 5), job(status = "running"), selecting = true))
    }

    @Test
    fun `statusAction running job offers pause`() {
        val action = statusAction(game(), job(status = "running"), selecting = false)
        assertEquals(StatusActionType.PAUSE, action?.type)
    }

    @Test
    fun `statusAction paused job offers resume`() {
        val action = statusAction(game(), job(status = "paused"), selecting = false)
        assertEquals(StatusActionType.RESUME, action?.type)
    }

    @Test
    fun `statusAction not-cached offers download`() {
        val action = statusAction(game(status = "idle"), null, selecting = false)
        assertEquals(StatusActionType.DOWNLOAD, action?.type)
    }

    @Test
    fun `statusAction error offers retry -- a deliberate extension over the mockup`() {
        val action = statusAction(game(status = "error"), null, selecting = false)
        assertEquals(StatusActionType.RETRY, action?.type)
    }

    @Test
    fun `statusAction cached is never actionable -- no silent re-download`() {
        val action = statusAction(game(status = "done", sizeBytes = 5), null, selecting = false)
        assertNull(action)
    }

    // ---- isJobStateTransition -------------------------------------------

    @Test
    fun `isJobStateTransition true when the job disappears`() {
        assertTrue(isJobStateTransition(job(status = "running"), null))
    }

    @Test
    fun `isJobStateTransition true for a brand-new job row`() {
        assertTrue(isJobStateTransition(null, job(status = "running")))
    }

    @Test
    fun `isJobStateTransition false when status is unchanged`() {
        assertFalse(isJobStateTransition(job(status = "running"), job(status = "running")))
    }

    @Test
    fun `isJobStateTransition true when status actually changed`() {
        assertTrue(isJobStateTransition(job(status = "running"), job(status = "paused")))
    }
}
